### PR TITLE
grpc: support full set of Google credentials.

### PR DESCRIPTION
**Pull request by htuch**
This supports things like OAuth, GCE default creds, refresh tokens, etc.

Risk Level: Low
Testing: Minimal tests. grpc::Channel/CallCredentials are opaque and don't provide much in the way
  of inspection from tests (CC: @vjpai).

Signed-off-by: Harvey Tuch <htuch@google.com>

### CODE REVIEW

File: source/common/grpc/google_grpc_creds_impl.cc [hunk 1]
```diff
@@ -4,13 +4,83 @@
...
 
-/**
- * TODO: Create GoogleGrpcCredentialsFactory for each built-in credential type defined in Google
- * gRPC.
- */
+std::shared_ptr<grpc::ChannelCredentials>
+defaultSslChannelCredentials(const envoy::api::v2::core::GrpcService& grpc_service_config,
+                             bool allow_insecure) {
+  const auto& google_grpc = grpc_service_config.google_grpc();
+  std::shared_ptr<grpc::ChannelCredentials> creds = nullptr;
```
Comment (lizan) [inline source/common/grpc/google_grpc_creds_impl.cc]: bring L16-21 to the bottom of this function then you don't need variable `creds`.

File: source/common/grpc/google_grpc_creds_impl.cc [hunk 2]
```diff
@@ -4,13 +4,83 @@
...
+  }
+  return creds;
+}
+
+std::shared_ptr<grpc::ChannelCredentials>
+defaultChannelCredentials(const envoy::api::v2::core::GrpcService& grpc_service_config,
+                          bool allow_insecure) {
+  std::shared_ptr<grpc::ChannelCredentials> creds =
+      defaultSslChannelCredentials(grpc_service_config, allow_insecure);
+  std::shared_ptr<grpc::CallCredentials> call_creds = nullptr;
```
Comment (lizan) [inline source/common/grpc/google_grpc_creds_impl.cc]: nit: no need `= nullptr`

File: source/common/grpc/google_grpc_creds_impl.cc [hunk 3]
```diff
@@ -4,13 +4,83 @@
...
+  for (const auto& credential : grpc_service_config.google_grpc().call_credentials()) {
+    std::shared_ptr<grpc::CallCredentials> new_call_creds;
+    switch (credential.credential_specifier_case()) {
```
Comment (lizan) [inline source/common/grpc/google_grpc_creds_impl.cc]: extract this switch to a function?

File: source/common/grpc/google_grpc_creds_impl.h [hunk 4]
```diff
@@ -13,5 +13,13 @@ grpc::SslCredentialsOptions buildSslOptionsFromConfig(
 std::shared_ptr<grpc::ChannelCredentials>
 getGoogleGrpcChannelCredentials(const envoy::api::v2::core::GrpcService& grpc_service);
 
+std::shared_ptr<grpc::ChannelCredentials>
+defaultSslChannelCredentials(const envoy::api::v2::core::GrpcService& grpc_service_config,
+                             bool allow_insecure = true);
```
Comment (lizan) [inline source/common/grpc/google_grpc_creds_impl.h]: I don't feel `allow_insecure = true` is right, make it `false` or no default?

File: test/common/grpc/google_grpc_creds_test.cc [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+
+namespace Envoy {
+namespace Grpc {
+namespace {
+
+// Force execution of all paths in Grpc::defaultChannelCredentials(). Since the underlying
+// grpc::{CallCredentials,ChannelCredentials} don't have any real way of getting at the underlying
+// state, we can at best just make sure we don't crash.
+TEST(DefaultChannelCredentials, All) {
+  { EXPECT_NE(nullptr, defaultChannelCredentials({})); }
```
Comment (lizan) [inline test/common/grpc/google_grpc_creds_test.cc]: since default is `true` (before my comment above) these 2 lines are testing same.

**Comment by lizan** [comment]
Also, remove corresponding `[#not-implemented-hide:]` from grpc_service.proto.

File: source/common/grpc/google_grpc_creds_impl.h (hunk 4 above)
Comment (htuch) [reply source/common/grpc/google_grpc_creds_impl.h]: This PR (other than the new creds stuff) is actually mostly just code movement from what was there already. I agree with your comments and will make said changes.

Comment (wozz) [reply source/common/grpc/google_grpc_creds_impl.h]: originally, this was set to true to remain consistent with the way it was implemented previously. I agree that call credentials shouldn't be enabled with an insecure channel though, so it makes sense to change it.

File: source/common/grpc/google_grpc_creds_impl.cc [hunk 6]
```diff
@@ -4,13 +4,83 @@
...
+    return grpc::SslCredentials(ssl_credentials_options);
+  }
+  return creds;
+}
+
+std::shared_ptr<grpc::ChannelCredentials>
+defaultChannelCredentials(const envoy::api::v2::core::GrpcService& grpc_service_config,
+                          bool allow_insecure) {
+  std::shared_ptr<grpc::ChannelCredentials> creds =
+      defaultSslChannelCredentials(grpc_service_config, allow_insecure);
```
Comment (wozz) [inline source/common/grpc/google_grpc_creds_impl.cc]: return this if allow_insecure is true? I don't think the default implementation should allow call credentials to be enabled on an insecure channel.

I think grpc may drop the call credentials on an insecure channel anyway: https://github.com/grpc/grpc/issues/8142

**Comment by lizan** [comment]
> return this if allow_insecure is true? I don't think the default implementation should allow call credentials to be enabled on an insecure channel.
>
> I think grpc may drop the call credentials on an insecure channel anyway: https://github.com/grpc/grpc/issues/8142

Let's not take the (about to change) restriction from gRPC library to here. We have discussed the use cases with gRPC security (cc @jiangtaoli2016) to and they are going to relax this on gRPC side. The particular use case we have is by sending credentials over unencrypted unix domain sockets, to bootstrap SDS.

cc @wattli @incfly

**Pull request merged**